### PR TITLE
Check arg type against Py UDF signature at query compile time

### LINKED ISSUE #5221: Move arg type check for Py UDF to query compile time
As we move to a more typed (early and strong) approach wrt Py UDF calls for mostly performance reasons, this is a first step towards the goal. We might also try to support numba decorated/numpy ufunc multi-signatures

related to #5112

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #5221

### CODE REVIEW

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java [hunk 1]
```diff
@@ -1968,6 +1968,39 @@ static boolean isWideningPrimitiveConversion(Class<?> original, Class<?> target)
         return false;
     }
 
+    public static boolean isLosslessWideningPrimitiveConversion(Class<?> original, Class<?> target) {
```
Comment (nbauernfeind) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:1971]: This is interesting. Historically we've allowed widening from Long to Float. I see how this is lossy.

Is this used to disable vectorization? Or is it used to disable lossy widening?

Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:1971]: See 5.1.2 on https://docs.oracle.com/javase/specs/jls/se9/html/jls-5.html. The intention here is to ensure lossless conversion.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java [hunk 2]
```diff
@@ -213,6 +239,44 @@ public void parseSignature() {
...
+        return false;
+    }
+
+
+    public void verifyArguments(Class<?>[] argTypes) {
+        String callableName = pyCallable.getAttribute("__name__").toString();
+
+        for (int i = 0; i < argTypes.length; i++) {
+            Set<Class<?>> types =
+                    parameters.get(i > parameters.size() - 1 ? parameters.size() - 1 : i).getPossibleTypes();
```
Comment (chipkent) [inline engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:260]: I'm not following what is happening here with the ternary indexing.  Is this some sort of vararg handling where the last parameter is presumed to be a vararg?  At a minimum, this deserves a comment.  At a maximum, the last parameter would indicate if it is a vararg parameter.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java [hunk 3]
```diff
@@ -213,6 +239,44 @@ public void parseSignature() {
...
+
+            // to prevent the unpacking of an array column when calling a Python function, we prefix the column accessor
+            // with a cast to generic Object type, until we can find a way to convey that info, we'll just skip the
+            // check for Object type input
+            if (argTypes[i] == Object.class) {
+                continue;
+            }
+
+            Class<?> t = getUnboxedType(argTypes[i]) == null ? argTypes[i] : getUnboxedType(argTypes[i]);
+            if (!types.contains(t) && !types.contains(Object.class)
```
Comment (chipkent) [inline engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:270]: Do you need a `PyObject` check?

File: py/server/deephaven/_udf.py [hunk 4]
```diff
@@ -92,14 +95,100 @@ def _encode_param_type(t: type) -> str:
...
+        if isinstance(t, np._typing._generic_alias._GenericAlias) and t.__origin__ == np.ndarray:
+            component_type = t.__args__[1].__args__[0]
+    # Py3.9+, np.ndarray as a generic alias is only supported in Python 3.9+, also npt.NDArray is still available but a
+    # specific alias (e.g. npt.NDArray[np.int64]) now is an instance of typing.GenericAlias.
+    # when npt.NDArray is used, the 1st argument is typing.Any, the 2nd argument is another generic alias of which
+    # the 1st argument is the component type
+    # when np.ndarray is used, the 1st argument is the component type
+    if not component_type and sys.version_info.major == 3 and sys.version_info.minor > 8:
+        import types
+        if isinstance(t, types.GenericAlias) and t.__origin__ == np.ndarray:
```
Comment (chipkent) [inline py/server/deephaven/_udf.py:159]: I see that `issubclass(t.__origin__, Sequence)` has been removed.  We planned to `Sequence` support, correct?

File: py/server/deephaven/_udf.py [hunk 5]
```diff
@@ -92,14 +95,100 @@ def _encode_param_type(t: type) -> str:
...
+    """Return True if the type is a Union type"""
+    if sys.version_info.major == 3 and sys.version_info.minor >= 10:
+        import types
+        if isinstance(t, types.UnionType):
+            return True
+
+    if isinstance(t, _GenericAlias) and t.__origin__ == Union:
+        return True
+
+    return False
```
Comment (chipkent) [inline py/server/deephaven/_udf.py:181]: These could be combined for less branching.

File: py/server/deephaven/_udf.py [hunk 6]
```diff
@@ -92,14 +95,100 @@ def _encode_param_type(t: type) -> str:
...
+
+    return np_dtype.char
+
+
+def _component_np_dtype_char(t: type) -> Optional[str]:
+    """Returns the numpy dtype character code for the given type's component type if the type is a Sequence type or
+    numpy ndarray, otherwise return None. """
+    component_type = None
+
+    if not component_type and sys.version_info.major == 3 and sys.version_info.minor > 8:
```
Comment (chipkent) [inline py/server/deephaven/_udf.py:116]: `component_type` appears to always be `None` here.

File: py/server/deephaven/_udf.py [hunk 7]
```diff
@@ -414,7 +505,8 @@ def wrapper(*args, **kwargs):
         j_class = real_ret_dtype.qst_type.clazz()
 
     wrapper.return_type = j_class
-    wrapper.signature = sig_str_vectorization
+    # wrapper.signature = sig_str_vectorization
+    wrapper.signature = p_sig.encoded
```
Comment (chipkent) [inline py/server/deephaven/_udf.py:509]: Which is right?  Is vectorization working?

File: py/server/tests/test_numba_guvectorize.py [hunk 8]
```diff
@@ -66,7 +66,10 @@ def g(x, dummy, res):
             res[0] = min(x)
             res[1] = max(x)
 
-        t = empty_table(10).update(["X=i%3", "Y=i"]).group_by("X").update("Z=g(Y,dummy)")
+        # convert dummy to a Java array
+        # TODO this is a hack, we might want to add a helper function for QLP to call to get the type of a PyObject arg
+        j_array = dtypes.array(dtypes.int64, dummy)
+        t = empty_table(10).update(["X=i%3", "Y=i"]).group_by("X").update("Z=g(Y,j_array)")
```
Comment (chipkent) [inline py/server/tests/test_numba_guvectorize.py:72]: Ideally, this should work without the hack... but making it "just work" has some other problems.  Need to give this some thought.  In the multidim array world, this could do a zero copy buffer handover.  Without that bit, there could be various bad outcomes.

File: py/server/tests/test_numba_guvectorize.py [hunk 9]
```diff
@@ -78,7 +81,11 @@ def g(x, dummy, res):
             res[0] = np.min(x)
             res[1] = np.max(x)
 
-        t = empty_table(10).update(["X=i%3", "Y=ii"]).group_by("X").update("Z=g(Y,dummy)")
+        # convert dummy to a Java array
+        # TODO this is a hack, we might want to add a helper function for QLP to call to get the type of a PyObject arg
+        j_array = dtypes.array(dtypes.int64, dummy)
```
Comment (chipkent) [inline py/server/tests/test_numba_guvectorize.py:86]: See comment above.

File: py/server/tests/test_udf_args.py [hunk 10]
```diff
@@ -287,7 +287,7 @@ def f6(*args: np.int32, col2: np.ndarray[np.int32]) -> bool:
 
         with self.assertRaises(DHError) as cm:
             t1 = t.update(["X1 = f6(X, Y=null)"])
-        self.assertIn("not compatible with annotation", str(cm.exception))
+        self.assertIn("f6: Expected argument (col2) to be one of [class [I], got boolean", str(cm.exception))
```
Comment (chipkent) [inline py/server/tests/test_udf_args.py:290]: A user, especially a python user is going to have a hard time with an error with type strings like `[I]`.  Can the error be made more user accessible?

File: py/server/tests/test_udf_args.py [hunk 11]
```diff
@@ -405,9 +405,9 @@ def f(x: {p_type}) -> bool:  # note typing
                 t = empty_table(1).update(["X = i", f"Y = f(({p_type})X)"])
                 self.assertEqual(1, t.to_string(cols="Y").count("true"))
 
-
-        np_int_types = {"np.int8", "np.int16", "np.int32", "np.int64"}
-        for p_type in np_int_types:
+    def test_np_typehints(self):
+        widening_np_int_types = {"np.int32", "np.int64"}
```
Comment (chipkent) [inline py/server/tests/test_udf_args.py:409]: I see tons of numpy testing changes, but I don't see testing changes for widening with plain python types.  Make sure that those types are well covered in tests since we are now pushing users to them.

File: py/server/tests/test_udf_args.py [hunk 12]
```diff
@@ -428,5 +440,52 @@ def f(x: {p_type}) -> bool:  # note typing
...
+        with self.subTest("bytearray"):
+            def f(x: bytearray) -> bool:
+                return True
+
+            with self.assertRaises(DHError) as cm:
+                t = empty_table(1).update(["X = i", "Y = f(ii)"])
+            self.assertRegex(str(cm.exception), "f: Expect")
+
+            t = empty_table(1).update(["X = i", "Y = (byte)(ii % 128)"]).group_by("X").update(["Z = f(Y.toArray())"])
+            self.assertEqual(t.columns[2].data_type, dtypes.bool_)
```
Comment (chipkent) [inline py/server/tests/test_udf_args.py:487]: It seems like there should also be return value tests for `bytes`, `bytearray` and whatever other new types we are supporting.

File: py/server/tests/test_udf_return_java_values.py [hunk 13]
```diff
@@ -54,7 +54,8 @@ def test_array_return(self):
             "np.str_": dtypes.string_array,
             "np.uint16": dtypes.char_array,
         }
-        container_types = ["List", "Tuple", "list", "tuple", "Sequence", "np.ndarray"]
+        # container_types = ["List", "Tuple", "list", "tuple", "Sequence", "np.ndarray"]
+        container_types = ["list"]
```
Comment (chipkent) [inline py/server/tests/test_udf_return_java_values.py:58]: Which here??  It seems like `list` is not sufficient.

File: py/server/tests/test_udf_args.py [hunk 14]
```diff
@@ -428,5 +440,52 @@ def f(x: {p_type}) -> bool:  # note typing
...
+
+
```
Comment (chipkent) [inline py/server/tests/test_udf_args.py:489]: Do we need a test for numpy arrays with widening/narrowing to make sure that we blow up in both cases?

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java
Comment (chipkent) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:1]: We are still missing logging to the system and python console for super-ass-slow numpy cases.  I suspect system logging might go here.

File: py/server/deephaven/_udf.py
Comment (chipkent) [inline py/server/deephaven/_udf.py:1]: We are still missing logging to the system and python console for super-ass-slow numpy cases.  I suspect python console logging might go here.

File: py/server/tests/test_udf_return_java_values.py
Comment (chipkent) [inline py/server/tests/test_udf_return_java_values.py:1]: It seems potentially prudent to have a few unit tests for function composition to make sure it functions reasonably instead of blowing up in some weird way.  e.g.

```python
def f1(x: int) -> float:
    ...

def f2(x: float) -> int:
  ...

t = t.update("X = f2(f1(ii))")
```
I'm sure you can also dream up some funky cases we should be considering.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java [hunk 15]
```diff
@@ -1968,6 +1968,39 @@ static boolean isWideningPrimitiveConversion(Class<?> original, Class<?> target)
...
+        if (original == null || !original.isPrimitive() || target == null || !target.isPrimitive()
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:1972]: Not possible. DH always unboxes a boxed type for a column.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java [hunk 16]
```diff
@@ -1968,6 +1968,39 @@ static boolean isWideningPrimitiveConversion(Class<?> original, Class<?> target)
...
+        LanguageParserPrimitiveType originalEnum = LanguageParserPrimitiveType.getPrimitiveType(original);
+
+        switch (originalEnum) {
+            case BytePrimitive:
+                if (target == short.class)
+                    return true;
+            case ShortPrimitive:
+            case CharPrimitive:
+                if (target == int.class)
+                    return true;
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:1990]: Java char is unsigned int16, so it should be treated as int32.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java [hunk 17]
```diff
@@ -2726,11 +2782,10 @@ private void checkVectorizability(@NotNull final MethodCallExpr n,
             }
         }
 
-        List<Class<?>> paramTypes = pyCallableWrapper.getParamTypes();
-        if (paramTypes.size() != expressions.length) {
+        if (pyCallableWrapper.getNumParameters() != expressions.length) {
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:2785]: Yes. It is only that vectorization won't work for them.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java [hunk 18]
```diff
@@ -2739,10 +2794,9 @@ private void prepareVectorizationArgs(
             Expression[] expressions,
             Class<?>[] argTypes,
             PyCallableWrapper pyCallableWrapper) {
-        List<Class<?>> paramTypes = pyCallableWrapper.getParamTypes();
-        if (paramTypes.size() != expressions.length) {
+        if (pyCallableWrapper.getNumParameters() != expressions.length) {
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:2797]: Yes. It is only that vectorization won't work for them.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java [hunk 19]
```diff
@@ -30,6 +30,7 @@ public class PyCallableWrapperJpyImpl implements PyCallableWrapper {
     private static final PyModule dh_udf_module = PyModule.importModule("deephaven._udf");
 
     private static final Map<Character, Class<?>> numpyType2JavaClass = new HashMap<>();
+    private static final Map<Character, Class<?>> numpyType2JavaArrayClass = new HashMap<>();
 
     static {
         numpyType2JavaClass.put('b', byte.class);
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:36]: Unsigned NP type is a little tricky. 

1. For input, only Java char matches uint16, other Java primitive types can't map to them.
2. For return value, wider Java types are required.

Therefore, I don't think we should support them except uint16.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java [hunk 20]
```diff
@@ -168,40 +183,51 @@ private void prepareSignature() {
...
+                Set<Class<?>> possibleTypes = new HashSet<>();
+                for (int ti = 0; ti < paramTypeCodes.length(); ti++) {
+                    char typeCode = paramTypeCodes.charAt(ti);
+                    if (typeCode == '[') {
+                        // skip the array type code
+                        ti++;
+                        possibleTypes.add(numpyType2JavaArrayClass.get(paramTypeCodes.charAt(ti)));
+                        if (paramTypeCodes.charAt(ti) == '?') {
+                            possibleTypes.add(Boolean[].class);
+                        }
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:216]: I actually don't think we support boolean[] in the engine because it doesn't allow null values.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java (hunk 3 above)
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:270]: No, No special handling of PyObject. 
1. It can't be used to type hint a parameter in a UDF
2. if it is a type of a column argument, the receiving parameter must be of Py/numpy object type. The same as other  unsupported Java types.

File: py/server/deephaven/_udf.py (hunk 4 above)
Comment (jmao-denver) [reply py/server/deephaven/_udf.py:159]: Yes, it is just moved to the calling func.

File: py/server/tests/test_numba_guvectorize.py (hunk 8 above)
Comment (jmao-denver) [reply py/server/tests/test_numba_guvectorize.py:72]: https://github.com/deephaven/deephaven-core/issues/5278

File: py/server/tests/test_numba_guvectorize.py (hunk 9 above)
Comment (jmao-denver) [reply py/server/tests/test_numba_guvectorize.py:86]: https://github.com/deephaven/deephaven-core/issues/5278

File: py/server/tests/test_udf_args.py (hunk 10 above)
Comment (jmao-denver) [reply py/server/tests/test_udf_args.py:290]: I think we can but it won't be a straightforward/non-trivial effort. Filed https://github.com/deephaven/deephaven-core/issues/5279

File: py/server/tests/test_udf_args.py (hunk 14 above)
Comment (jmao-denver) [reply py/server/tests/test_udf_args.py:489]: Probably not as the code logic guarantees each argument will be checked.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:1]: It will be in the follow up PR where we use specific type converters.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (chipkent) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java]: I'm asking why it is even considered an int not why is it an int32.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (chipkent) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java]: Do we need to support the boxed version of anything else?

[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java]: No, these are automatically unboxed by the engine.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java [hunk 21]
```diff
@@ -2498,13 +2499,36 @@ public Class<?> visit(MethodCallExpr n, VisitArgs printer) {
...
                 typeArguments);
     }
 
+    private void verifyPyCallableArguments(@NotNull MethodCallExpr n, @NotNull Class<?>[] argTypes) {
+        final String invokedMethodName = n.getNameAsString();
+
+        if (GET_ATTRIBUTE_METHOD_NAME.equals(invokedMethodName)) {
+            // Only PyCallableWrapper.getAttribute()/PyCallableWrapper.call() may be invoked from the query language.
+            // UDF type checks are not currently supported for getAttribute() calls.
+            return;
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:2516]: Right, and we have a ticket to add support of them. #3121

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java [hunk 22]
```diff
@@ -168,48 +177,124 @@ private void prepareSignature() {
...
+            throw new IllegalArgumentException("Arguments must be a primitive type (excluding void)!");
+        }
+
+        if (original.equals(target)) {
+            return true;
+        }
+
+        if (original.equals(byte.class)) {
+            return target == short.class || target == int.class || target == long.class;
+        } else if (original.equals(short.class) || original.equals(char.class)) { // char is unsigned, so it's a
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:261]: The 'for' argument is that Java allows it and it is a lossless coercion.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java]: I think my answer implies that Java char is considered to be an integral type, which is probably the common consensus out there.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (chipkent) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:2516]: I would reword it to be a little more clear.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java [hunk 23]
```diff
@@ -168,48 +177,124 @@ private void prepareSignature() {
...
+                return true;
+            }
+            if (t.isPrimitive() && type.isPrimitive() && isLosslessWideningPrimitiveConversion(type, t)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static boolean isLosslessWideningPrimitiveConversion(Class<?> original, Class<?> target) {
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:249]: 1. This particular class is not being tested in the JUnit tests probably due to the fact that it needs a non-trivial amount of work to properly wire everything up to test it in Java.
2. The Python unit tests indirectly test this method in a pretty comprehensive way.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java [hunk 24]
```diff
@@ -168,48 +177,124 @@ private void prepareSignature() {
...
+    }
+
+    public void verifyArguments(Class<?>[] argTypes) {
+        String callableName = pyCallable.getAttribute("__name__").toString();
+        List<Parameter> parameters = signature.getParameters();
+
+        for (int i = 0; i < argTypes.length; i++) {
+            // if there are more arguments than parameters, we'll need to consider the last parameter as a varargs
+            // parameter. This is not ideal. We should consider a better way to handle this, i.e. a way to convey that
+            // the function is variadic.
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:280]: Yes, the signature include both. We do have a variadic test.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (chipkent) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:261]: I think the link included in a prior comment is sufficient to include it.  https://docs.oracle.com/javase/specs/jls/se9/html/jls-5.html#jls-5.1.2

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (chipkent) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java]: I think https://docs.oracle.com/javase/specs/jls/se9/html/jls-5.html#jls-5.1.2 is sufficient justification.

File: engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (chipkent) [reply engine/table/src/main/java/io/deephaven/engine/util/PyCallableWrapperJpyImpl.java:249]: 1. This is a static method that takes `Class` parameters.  It should be very easy to unit test in Java.
2. Unit tests like this should happen in Java when possible.  That way things like coverage can be assessed.  
3. Additional Python tests should support the Java testing and not replace it.  
4. I suspect the lack of unit tests resulted from a lack of diligence at some point in the last 15 years.  Now we are at a point where it looks like a lot of work to test everything.  This is why I suggested testing a single static function.

**Review by chipkent — APPROVED** [review]
I'm approving but still think there should be a Java unit test as noted.

**Comment by deephaven-internal** [comment]
Labels indicate documentation is required. Issues for documentation have been opened:

Community: https://github.com/deephaven/deephaven-docs-community/issues/179
